source/generators/SingleParticle.cc

```diff
@@ -110,80 +107,48 @@ void SingleParticle::SetParticleDefinition(G4String particle_name)
 {
-  // Generate an initial position for the particle using the geometry
-  G4ThreeVector position = geom_->GenerateVertex(region_);
-
-  // Particle generated at start-of-event
-  G4double time = 0.;
-
-  // Create a new vertex
-  G4PrimaryVertex* vertex = new G4PrimaryVertex(position, time);
-
   // Generate uniform random energy in [E_min, E_max]
-  G4double kinetic_energy = RandomEnergy();
-
-  // Generate random direction by default
-  G4ThreeVector momentum_direction_ = G4RandomDirection();
+  G4double kinetic_energy = nexus::RandomEnergy(energy_max_, energy_min_);
 
-    // Calculate cartesian components of momentum
+  // Calculate cartesian components of momentum
   G4double mass   = particle_definition_->GetPDGMass();
   G4double energy = kinetic_energy + mass;
   G4double pmod = std::sqrt(energy*energy - mass*mass);
-  G4double px = pmod * momentum_direction_.x();
-  G4double py = pmod * momentum_direction_.y();
-  G4double pz = pmod * momentum_direction_.z();
-
-  // If user provides a momentum direction, this one is used
-  if (momentum_X_ != 0. || momentum_Y_ != 0. || momentum_Z_ != 0.) {
-    // Normalize if needed
-    G4double mom_mod = std::sqrt(momentum_X_ * momentum_X_ +
-				 momentum_Y_ * momentum_Y_ +
-				 momentum_Z_ * momentum_Z_);
-    px = pmod * momentum_X_/mom_mod;
-    py = pmod * momentum_Y_/mom_mod;
-    pz = pmod * momentum_Z_/mom_mod;
-  } else if (costheta_min_ != -1. || costheta_max_ != 1. || phi_min_ != 0. || phi_max_ !=2.*pi) {
-    G4bool mom_dir = false;
-    while (mom_dir == false) {
-      G4double cosTheta  = 2.*G4UniformRand()-1.;
-      if (cosTheta > costheta_min_ && cosTheta < costheta_max_){
-	G4double sinTheta2 = 1. - cosTheta*cosTheta;
-	if( sinTheta2 < 0.)  sinTheta2 = 0.;
-	G4double sinTheta  = std::sqrt(sinTheta2);
-	G4double phi = twopi*G4UniformRand();
-	  if (phi > phi_min_ && phi < phi_max_){
-	    mom_dir = true;
-	    momentum_direction_ = G4ThreeVector(sinTheta*std::cos(phi),
-						sinTheta*std::sin(phi), cosTheta).unit();
-	    px = pmod * momentum_direction_.x();
-	    py = pmod * momentum_direction_.y();
-	    pz = pmod * momentum_direction_.z();
-	  }
-      }
-    }
+
+  bool fixed_momentum = momentum_ != G4ThreeVector{};
+  bool restrict_angle = costheta_min_ != -1. || costheta_max_ != 1. || phi_min_ != 0. || phi_max_ !=2.*pi;
+
+  G4ThreeVector p_dir; // it will be set in the if branches below
+  if (fixed_momentum) { // if the user provides a momentum direction
+    p_dir = momentum_.unit();
+  } else if (restrict_angle) { // if the user provides a range of angles
+    p_dir = nexus::Direction(costheta_min_, costheta_max_, phi_min_, phi_max_);
+  } else {
+    p_dir = G4RandomDirection();
   }
 
+  G4ThreeVector p = pmod * p_dir;
+
   // Create the new primary particle and set it some properties
   G4PrimaryParticle* particle =
-    new G4PrimaryParticle(particle_definition_, px, py, pz);
+    new G4PrimaryParticle(particle_definition_, p.x(), p.y(), p.z());
```

Comment:
Use `auto` to shorten this to a single line.

Reply:
Done.

Reply:
I think that marking the conversation as resolved is probably enough, without having to add a "Done." comment: the arrival of the corresponding commit should be enough to attract attention.